aten/src/ATen/native/cuda/Copy.cu

```diff
@@ -0,0 +1,245 @@
+struct CopyOp {
+  static void apply(Tensor& dst, const Tensor& src) {
+    CUDA_tensor_apply2<dst_T, src_T>(
+        dst, src, [] __device__(dst_T & dst_val, const src_T& src_val) {
+#if __CUDA_ARCH__ >= 350
+          dst_val = static_cast<dst_T>(
+              static_cast<native::inter_copy_type_t<dst_T>>(__ldg(&src_val)));
+#else
+          dst_val = static_cast<dst_T>(static_cast<native::inter_copy_type_t<dst_T>>(src_val));
+#endif
+        });
+  }
+};
+
+// device-to-device copy, does type conversion
+template <typename dst_T, typename src_T>
+void copy_device_to_device(Tensor& dst, const Tensor& src) {
+  auto numel = dst.numel();
+  if (dst.is_same(src) || numel == 0) {
+    return;
+  }
+
+  // We can memcpy the memory if:
+  // -both tensors are contiguous; or,
+  // -there is only one element to copy; or,
+  // -FIXME: if both tensors have matching size and stride arrays, and no
+  // holes within (in other words, there is some permutation that can be applied
+  // to the size/strides such that the resulting tensor is
+  // contiguous).
+  // -AND: both tensors have the same type.
+  bool same_type = std::is_same<dst_T, src_T>::value;
+  bool memcpy_eligible =
+      ((src.is_contiguous() && dst.is_contiguous()) || (numel == 1)) &&
+      same_type;
+
+  Device src_device = src.device();
+  Device dst_device = dst.device();
+
+  // Try to enable p2p access. This also handles the case src_device ==
+  // dst_device.
+  bool p2pEnabled = THCState_getPeerToPeerAccess(
+      globalContext().getTHCState(), src_device.index(), dst_device.index());
+
+  // We always perform the copy on the source device, using the
+  // current stream on the source device.
+  // If the copy is on the default stream, then we fully synchronize
+  // both src and dst's default streams for completion of the
+  // copy. We have to explicitly do this for non-contig copies.
+  // This mimics the behavior of cross-device cudaMemcpyAsync on
+  // the default stream.
+  // If the copy is not on the default stream, then it is up to the
+  // user to add needed synchronization on the dst device, since the
+  // stream on the dst device that wishes to synchronize may not be
+  // the same index as the one on the src device.
+  CUDAStream copy_stream = getCurrentCUDAStream(src_device.index());
+  if (src_device != dst_device && copy_stream == NULL) {
+    // This is a cross-device copy on the default stream. We perform a
+    // two-way barrier between both devices' default streams before
+    // the copy. This ensures that any write-after-write and
+    // write-after-read dependencies on the destination side are
+    // handled, so that no one is operating on the dst memory when
+    // we perform the copy.
+    // src waits on dst barrier (src already waits on src)
+    CUDAEvent dst_ready;
+    DeviceGuard device_guard_dst{dst_device};
+    dst_ready.record(getDefaultCUDAStream(dst_device.index()));
+
+    DeviceGuard device_guard_src{src_device};
```

Comment:
Then you can use `set_device` here to skip the unnecessary reset of the first device guard.